File: 198. House Robber/solution.go
func rob(nums []int) int {
    length := len(nums)
    if length == 0 {
        return 0
    }
    if length == 1 {
        return nums[0]
    }
    a := nums[0]
    b := nums[1]
    record := []int{a, max(a, b)}
    for i:=2; i<length; i++ {
        record = append(record, max(record[len(record)-1], record[len(record)-2] + nums[i]))
    }
    return record[len(record)-1]
}

func max(a int, b int) int {
    if a >= b {
        return a
    }
    return b
}

// 一开始思路是从0开始，或者从1开始，比较每次跳一个或两个，取最大，最后两个出发点取最大
// 反例：[2,4,8,9,9,3]
// 应该使用贪心算法

File: 198. House Robber/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    if(nums.length === 0) return 0
    if(nums.length === 1) return nums[0]
    if(nums.length === 2) return Math.max(nums[0], nums[1])
    
    let arr = [nums[0], Math.max(nums[0], nums[1])] //局部最优 -> 全局最优，arr存储偷取每一家的最优解
    for(let i=2; i<nums.length; i++) {
        arr.push(Math.max(nums[i] + arr[i-2], arr[i-1])) // 基于上一步的最优解，比较偷当前house，不偷当前house，选择最大值
    }
    return arr[arr.length-1]
};